### PR TITLE
[BOO] Enable Backward Convolution in Experimental Schema

### PR DESCRIPTION
**Pull request by zjgarvey**
There are several changes in this PR:

### Core Turbine Changes

1. `attr_list_bool` is added as a method for `KernelSelection`, and child classes are updated to support this.
2. `maybe_return_tensor` is added as a method for `KernelSelection`, which returns either a `TensorArg` or `EmptyOptionalTensorArg`.
3. `KernelBuilder.yield_results` is modified for `FreeFuncKernelBuilder` and `InlineKernelBuilder` to manage `None` type returns for optional return tensors.
4. Eager execution for `CustomOp` is modified to properly return `None` for non-returned outputs.
5. The aot export process will now better report when a `None` type is found on the function boundary of an exported program, rather than hitting a `KeyError`.

### BOO changes

1. `Mode` for conv now includes all combinations of backward modes. Mode values (`--forw`) not included in the original miopen driver command are included in the parser for testing.
2. `ConvSignature` can now be initialized with an optional `backward_mask` kwarg, which will infer the mode based on the provided mask.
3. All backward `nn.Module` implementations for backward convolution now take all three `(dLdy, x, w)` regardless of whether an input is redundant or not. This design decision is based on the fact that almost every real instance of backward conv will need both, and it greatly simplifies the various backward modules. For standalone convolutions which only require weight/bias gradient may suffer some cpu overhead from exporting the unused weight tensor, so this decision should probably be discussed.
4. `ConvSignature.get_nn_module` supports all backward combination modes. When used with `use_custom=False`, the module simply handles layout transformations and applies `aten.convolution_backward`. When used with `use_custom=True`, the module will use existing custom submodules for each grad computation.
5. All implementations of `boo.convolution_backward` now use combined backward modes instead of separating into multiple launchables. We can modify this if necessary, but there is non-trivial CPU overhead involved in launching two separate vmfbs that we should be able to mitigate by merging.
6. All implementations of `boo.convolution_backward` now use a signature that more closely aligns with `aten.convolution_backward`.
7. The custom operator `boo.layout_customizable_convolution_backward` now supports IR generation using the core infra changes mentioned above.
8. A replacement function is added for convolution backward, and the backward conv op and this replacement function are included in the relevant experimental schemas.

### CODE REVIEW

File: iree/turbine/transforms/general/custom_op_expansion.py [hunk 1]
```diff
@@ -305,6 +326,31 @@ def _get_constant_str_from_value(v: Value) -> str:
     return StringAttr(constant_op.attributes["value"]).value
 
 
+def _get_constant_bool_from_value(v: Value) -> bool:
+    """Given a constant bool producer, return the bool.
+
+    Example: %bool = torch.constant.bool False
+    """
+    constant_op = OpResult(v).owner
```
[the remark this answers is not in the document]
Comment (zjgarvey) [reply iree/turbine/transforms/general/custom_op_expansion.py:334]: I suppose that's a good point. I don't think we expect any `!torch.bool` block arguments at this level, but it's theoretically possible.

[the remark this answers is not in the document]
Comment (zjgarvey) [reply iree/turbine/transforms/general/custom_op_expansion.py:334]: Ah, this is only used for things we expect to be constants, so they should never be block arguments. This should throw at some point if that's not the case.
